### ISSUE TITLE
Check for Source IP/CIDR in Policies

### ISSUE THREAD

**Issue by meshuga** [leandrodamascena/aws-network-discovery#14]
It's possible to allow access based on `aws:SourceIp`keyword [1].

The task is to detect that and try to correlate VPC with VPC's subnet.
Sine there is no 100% guarantee the correlation will be correct, an information about similarity must be displayed, if potential subnets are detected within account.

Refs
[1] https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_condition-keys.html#condition-keys-sourceip

**Comment by leandrodamascena** [comment]
Imagine that I have this policy..

"Condition": {
                "Bool": {"aws:ViaAWSService": "false"},
                "IpAddress": {"aws:SourceIp": "123.45.167.89"}
            }

What you're saying is when found key ipAddress-SourceIp, find in all subnets from that vpc to discover if its a potencial Ip used? Makes sense. This will require a little more attention cause some ips are a CIDR block and not explict IP.

Good.

**Comment by meshuga** [comment]
Yes, that's what I had in mind. There are ofc some pitfalls. If an IP address/CIDR is within a range of overlapping CIDRs like in [1], things aren't getting so obvious.

[1] https://docs.aws.amazon.com/vpc/latest/peering/peering-configurations-partial-access.html#one-to-two-vpcs-simple-hub

**Comment by leandrodamascena** [comment]
Ok, I get it! I'll simulate some scenarios.

**Comment by leandrodamascena** [comment]
@meshuga it's a prototype in a separated branch... There are some issues to fix before merge in developer branch:

1 - Works only if the key "aws:SourceIp" is a single IP, not a list.
2 - Works only if the key "aws:SourceIp" has a IP with no subnet block.. Example: 10.0.1.2 works.. 10.0.1.2/32 doens't work.

I need to improve reg expression to fix these itens.. But it's working :D

![image](https://user-images.githubusercontent.com/4295173/81169443-c87e3b80-8f90-11ea-98e6-c2335ffdee32.png)

![image](https://user-images.githubusercontent.com/4295173/81169516-e0ee5600-8f90-11ea-8a88-e289ea5912a1.png)

![image](https://user-images.githubusercontent.com/4295173/81169548-eb105480-8f90-11ea-91c1-587da01cfef7.png)

![image](https://user-images.githubusercontent.com/4295173/81169606-024f4200-8f91-11ea-8578-74f9b6be9626.png)

**Comment by leandrodamascena** [comment]
Maybe you'll need to run: pip install ipaddress

**Comment by leandrodamascena** [comment]
Fixed

> 2 - Works only if the key "aws:SourceIp" has a IP with no subnet block.. Example: 10.0.1.2 works.. 10.0.1.2/32 doens't work.

**Comment by leandrodamascena** [comment]
![image](https://user-images.githubusercontent.com/4295173/81187411-eefda000-8fab-11ea-8a09-78e162aec5ca.png)

![image](https://user-images.githubusercontent.com/4295173/81187480-06d52400-8fac-11ea-988c-a283138d42fb.png)

**Comment by meshuga** [comment]
That policy doesn't much work (nothing is detected):
```
{
  "Version": "2012-10-17",
  "Id": "arn:aws:sqs:ap-southeast-1:xxx:aaaaa/SQSDefaultPolicy",
  "Statement": [
    {
      "Sid": "xxx",
      "Effect": "Allow",
      "Principal": "*",
      "Action": "SQS:*",
      "Resource": "arn:aws:sqs:ap-southeast-1:xxx:aaaaa",
      "Condition": {
        "IpAddress": {
          "aws:SourceIp": [
            "127.0.0.1",
            "192.168.0.1",
            "172.1.1.0/24",
            "2001:DB8:1234:5678::/64"
          ]
        }
      .
    }
  ]
}
```
Policy without IPv6 CIDR is not detected as well.

I got an idea: to make the tool be solid - let's just add another detection mechanism and not correlate it with any VPC or subnet whatsoever. Let's just display the information as is. It will reduce potential misunderstanding of the report entries.

**Comment by leandrodamascena** [comment]
Is it a IAM Policy or SQS Policy inside a queue? In this branch the tool just check if policy is inside the queue not just as IAM Policy. I'll replicate if success in SQS Queue Policy..

![image](https://user-images.githubusercontent.com/4295173/81195567-eb6f1680-8fb5-11ea-83da-54679f72f8fd.png)

And about your ideia, I didn't get it... explain more, please.

**Comment by meshuga** [comment]
Yup, it's an access policy assigned to my queue, as shown below:
![image](https://user-images.githubusercontent.com/1073936/81197058-0a879b80-8fea-11ea-9675-8a617414429e.png)

**Comment by leandrodamascena** [comment]
![image](https://user-images.githubusercontent.com/4295173/81197561-591c4200-8fb8-11ea-9eab-02b0db5d11fe.png)

![image](https://user-images.githubusercontent.com/4295173/81197577-5e798c80-8fb8-11ea-9dd7-e43c3ee01ddc.png)

Maybe yours subnets CIDR? I can enable except too and we can check if some error shows..

And tell me more about your ideia.

**Comment by meshuga** [comment]
I pointed that out in https://github.com/leandrodamascena/aws-network-discovery/issues/14#issuecomment-624711033
The idea is not to display network information from a context of solely VPCs but also other, like in this case, IP addresses, subnets.

**Comment by leandrodamascena** [comment]
Ok, get it! I'll able debug in this version and you check if some error...

**Comment by leandrodamascena** [comment]
Try now... script will show exceptions when check if a ip is inside a subnet..

**Comment by leandrodamascena** [comment]
Merged this branch into developer.

**Comment by leandrodamascena** [comment]
Included in 0.5.0 release.
